File: CodingInterviewsPackSolutionsCPP/LFU Cache.cpp
class LFUCache {
    int capacity{};
    int minFreq{};
    
    unordered_map<int,int>keyValue;
    unordered_map<int,int>keyFreq;
    unordered_map<int,list<int>>freqKeylist;
    unordered_map<int,list<int>::iterator>keyIterator;
public:
    LFUCache(int capacity):capacity(capacity) {
        
    }
    
    int get(int key) {
        if(!keyValue.count(key))
            return -1;
        
        updateLRU(key);
        return keyValue[key];
    }
    
    void put(int key, int value) {
        if(capacity==0)
            return;
        if(capacity==keyValue.size() && keyValue.count(key)==0)
            evict();
        if(keyValue.count(key)){
            updateLRU(key);
            keyValue[key] = value;
        }
        else{
            keyValue[key] = value;
            keyFreq[key] = 1;
            freqKeylist[1].push_front(key);
            keyIterator[key] = freqKeylist[1].begin();
            minFreq = 1;
        }
    }
    
    void updateLRU(int key){
        freqKeylist[keyFreq[key]].erase(keyIterator[key]);
        keyFreq[key]++;
        freqKeylist[keyFreq[key]].push_front(key);
        keyIterator[key] = freqKeylist[keyFreq[key]].begin();
        
        if(freqKeylist[minFreq].empty()){
            freqKeylist.erase(minFreq);
            minFreq++;
        }
    }
    
    void evict(){
        int lfu = freqKeylist[minFreq].back();
        freqKeylist[minFreq].pop_back();
        keyValue.erase(lfu);
        keyFreq.erase(lfu);
        keyIterator.erase(lfu);
    }
};

File: CodingInterviewsJavaSolutions/LFU Cache.java
class LFUCache {
    private int capacity;
    private int minFreq;
    private Map<Integer, Integer> keyValue;
    private Map<Integer, Integer> keyFreq;
    private Map<Integer, List<Integer>> freqKeylist;
    private Map<Integer, Integer> keyIterator;

    public LFUCache(int capacity) {
        this.capacity = capacity;
        this.minFreq = 0;
        this.keyValue = new HashMap<>();
        this.keyFreq = new HashMap<>();
        this.freqKeylist = new HashMap<>();
        this.keyIterator = new HashMap<>();
    }

    public int get(int key) {
        if (!keyValue.containsKey(key)) {
            return -1;
        }
        updateLRU(key);
        return keyValue.get(key);
    }

    public void put(int key, int value) {
        if (capacity == 0) {
            return;
        }
        if (capacity == keyValue.size() && !keyValue.containsKey(key)) {
            evict();
        }
        if (keyValue.containsKey(key)) {
            updateLRU(key);
            keyValue.put(key, value);
        } else {
            keyValue.put(key, value);
            keyFreq.put(key, 1);
            freqKeylist.putIfAbsent(1, new LinkedList<>());
            freqKeylist.get(1).add(0, key);
            keyIterator.put(key, freqKeylist.get(1).get(0));
            minFreq = 1;
        }
    }

    private void updateLRU(int key) {
        freqKeylist.get(keyFreq.get(key)).remove(Integer.valueOf(keyIterator.get(key)));
        keyFreq.put(key, keyFreq.get(key) + 1);
        int freq = keyFreq.get(key);
        freqKeylist.putIfAbsent(freq, new LinkedList<>());
        freqKeylist.get(freq).add(0, key);
        keyIterator.put(key, freqKeylist.get(freq).get(0));

        if (freqKeylist.get(minFreq).isEmpty()) {
            freqKeylist.remove(minFreq);
            minFreq++;
        }
    }

    private void evict() {
        int lfu = freqKeylist.get(minFreq).remove(freqKeylist.get(minFreq).size() - 1);
        keyValue.remove(lfu);
        keyFreq.remove(lfu);
        keyIterator.remove(lfu);
    }
}
